contracts/button/button.rs

```diff
@@ -0,0 +1,399 @@
+                "Contract call failed due to NotCallable error",
+            )),
+            InkEnvError::Unknown => {
+                Error::ContractCall(String::from("Contract call failed due to Unknown error"))
+            }
+            InkEnvError::LoggingDisabled => Error::ContractCall(String::from(
+                "Contract call failed due to LoggingDisabled error",
+            )),
+            InkEnvError::EcdsaRecoveryFailed => Error::ContractCall(String::from(
+                "Contract call failed due to EcdsaRecoveryFailed error",
+            )),
+            #[cfg(any(feature = "std", test, doc))]
+            InkEnvError::OffChain(_e) => {
+                Error::ContractCall(String::from("Contract call failed due to OffChain error"))
+            }
+        }
+    }
+}
+
+/// Game contracts storage
+#[derive(Debug, SpreadLayout, SpreadAllocate)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink_storage::traits::StorageLayout)
+)]
+pub struct ButtonData {
+    /// How long does TheButton live for?
+    pub button_lifetime: u32,
+    /// Stores a mapping between user accounts and the number of blocks they extended The Buttons life for
+    pub presses: Mapping<AccountId, u32>,
+    /// stores keys to `presses` because Mapping is not an Iterator. Heap-allocated so we might need Map<u32, AccountId> if it grows out of proportion
+    pub press_accounts: Vec<AccountId>,
+    /// stores total sum of user scores
+    pub total_scores: u32,
+    /// stores the last account that pressed The Button
+    pub last_presser: Option<AccountId>,
+    /// block number of the last press
+    pub last_press: u32,
+    /// AccountId of the ERC20 ButtonToken instance on-chain
+    pub button_token: AccountId,
+    /// accounts whitelisted to play the game
+    pub can_play: Mapping<AccountId, bool>,
+    /// access control contract
+    pub access_control: AccountId,
+}
+
+/// Concrete implementations of the games API
+///
+/// Implementing contract needs to return ButtonData read-only and mutably
+/// as well as implement `score` the logic that based on the current block number and the contract storage state returns a users score.
+/// Remaining methods have default implementations that can be overriden as needed.
+///
+/// NOTE: no contract events are being emitted, so the implementing contract is responsible for defining and emitting those.
+pub trait ButtonGame {
+    /// Getter for the button data
+    fn get(&self) -> &ButtonData;
+
+    /// Mutable getter for the button data
+    fn get_mut(&mut self) -> &mut ButtonData;
+
+    /// Logic for calculating user score given the particular games rules
+    fn score(&self, now: u32) -> u32;
+
+    fn is_dead(&self, now: u32) -> bool {
+        now > self.deadline()
+    }
+
+    fn deadline(&self) -> u32 {
+        let ButtonData {
+            last_press,
+            button_lifetime,
+            ..
+        } = self.get();
+        last_press + button_lifetime
+    }
```

Comment:
I'm just wondering where is the moment we would like to keep such information in storage instead of computing it every time

do you have some rule of thumb?

Reply:
I suppose storage >> computation, so when computing is cheap like here :shrug: